File: 0046-permutations/0046-permutations.py
class Solution:
    def permute(self, nums: List[int]) -> List[List[int]]:
        n = len(nums)
        result = []

        def dfs(path):
            if len(path) == n:
                result.append(path[:])
                return
            
            for i in range(n):
                if nums[i] in path:
                    continue
                
                path.append(nums[i])
                dfs(path)
                path.pop()
                
        
        dfs([])

        return result

File: 0046-permutations/0046-permutations.cpp
class Solution {
public:
    vector<vector<int>> answer;
    vector<int> path;

    void dfs(int n, vector<int>& path, vector<int>& nums) {
        if (path.size() == n) {
            answer.push_back(path);
            return;
        }

        for (int i=0; i<n; i++) {
            if (find(path.begin(), path.end(), nums[i]) == path.end()) {
                path.push_back(nums[i]);
                dfs(n, path, nums);
                path.pop_back();
            }
        }
    }

    vector<vector<int>> permute(vector<int>& nums) {
        int n = nums.size();
        dfs(n, path, nums);
        return answer;
    }
};